[jrubyscripting] JRuby scripting initial contribution (#11538)
openhab/openhab-addons: bundles/org.openhab.automation.jrubyscripting/README.md

# JRuby Scripting 

This add-on provides [JRuby](https://www.jruby.org/) 9.3.1 that can be used as a scripting language within automation rules.

## JRuby Scripting Configuration

> JRuby configuration parameters may be set by creating a jruby.cfg file in $OPENHAB_CONF/services/

Reviewer: Is there a reason why it is only possible to configure this using a file and not using the UI? 

To make this configurable using the UI you can add a `ConfigurableService` like the openhabcloud [CloudService](https://github.com/openhab/openhab-addons/blob/c79fd062569221451e6809c4afd52ecbac9822a9/bundles/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/CloudService.java#L67-L71).  Such services get their configuration using config descriptions, very similar to bindings, see [config.xml](https://github.com/openhab/openhab-addons/blob/c79fd062569221451e6809c4afd52ecbac9822a9/bundles/org.openhab.io.openhabcloud/src/main/resources/OH-INF/config/config.xml).

Author: Is it possible to have default values that reference directories ($OPENHAB_CONF) in the UI? Or do we say what the default is in the description?  If we do the UI for this, do we get an empty string if they don't fill anything out or do we get ""?

Reviewer: It probably won't expand the variables in paths automatically, but you could make it an optional parameter and explain in the description how it behaves when it is left empty. That is the approach used for this in the [upnpcontrol binding](https://github.com/openhab/openhab-addons/blob/2a8e9b6e938479cfeb21a4ebd0251a6207c86030/bundles/org.openhab.binding.upnpcontrol/src/main/resources/OH-INF/binding/binding.xml#L9-L15).

Another approach is to expand the variable yourself which seems to be done in the [logreader binding](https://github.com/openhab/openhab-addons/blob/c79fd062569221451e6809c4afd52ecbac9822a9/bundles/org.openhab.binding.logreader/src/main/resources/OH-INF/thing/reader.xml#L27-L31).

I don't have a strong preference for one implementation over the other as long as it can be configured using the UI.

It might be a good idea to add support for expanding environment variables in strings in openhab-core one day though so it can be used anywhere without having to duplicate the code for this. :slightly_smiling_face:

| org.openhab.automation.jrubyscripting:local_context   | threadsafe                              | The local context holds Ruby runtime, name-value pairs for sharing variables between Java and Ruby. See [this](https://github.com/jruby/jruby/wiki/RedBridge#Context_Instance_Type) for options and details |
| org.openhab.automation.jrubyscripting:local_variables | transient                               | Defines how variables are shared between Ruby and Java. See [this](https://github.com/jruby/jruby/wiki/RedBridge#local-variable-behavior-options) for options and details                                   |
| org.openhab.automation.jrubyscripting:gems            |                                         | Comma separated list of [Ruby Gems](https://rubygems.org/) to install.                                                                                                                                      |

## Ruby Gems

> This binding will install user specified gems and make them available on the library search path.

Reviewer: ```suggestion
This automation add-on will install user specified gems and make them available on the library search path.
```

Gem versions may be specified using the standard ruby gem_name=version format.
The version number follows the [pessimistic version constraint](https://guides.rubygems.org/patterns/#pessimistic-version-constraint) syntax. 

> For example this configuration will install version 4 or higher of the [OpenHAB JRuby Scripting Libray](https://boc-tothefuture.github.io/openhab-jruby/).

Reviewer: ```suggestion
For example this configuration will install version 4 or higher of the [openHAB JRuby Scripting Libray](https://boc-tothefuture.github.io/openhab-jruby/).
```

```text
org.openhab.automation.jrubyscripting:gems=openhab-scripting=~>4.0
```

## Creating JRuby Scripts

When this add-on is installed, you can select JRuby as a scripting language when creating a script action within the rule editor of the UI.

Alternatively, you can create scripts in the `automation/jsr223` configuration directory.
If you create an empty file called `test.rb`, you will see a log line with information similar to:

```text
    ... [INFO ] [.a.m.s.r.i.l.ScriptFileWatcher:150  ] - Loading script 'test.rb'
```

To enable debug logging, use the [console logging]({{base}}/administration/logging.html) commands to
enable debug logging for the automation functionality:

```text
log:set DEBUG org.openhab.core.automation

> log:set DEBUG org.openhab.binding.jrubyscripting

Reviewer 2: ```suggestion
log:set DEBUG org.openhab.automation.jrubyscripting
```
